exla/livebooks/rotating_image.livemd

```diff
@@ -0,0 +1,206 @@
+
+    y =
+      orig_ctr_height
+      |> Nx.add(1)
+      |> Nx.add(i)
+      |> then(&Nx.subtract(y_image, &1))
+
+    x =
+      orig_ctr_width
+      |> Nx.add(1)
+      |> Nx.add(j)
+      |> then(&Nx.subtract(x_image, &1))
+
+    tangent =
+      angle
+      |> Nx.divide(2)
+      |> Nx.tan()
+
+    new_x =
+      y
+      |> Nx.multiply(tangent)
+      |> then(&Nx.subtract(x, &1))
+      |> Nx.round()
+
+    new_y =
+      angle
+      |> Nx.sin()
+      |> Nx.multiply(new_x)
+      |> Nx.add(y)
+      |> Nx.round()
+
+    new_x =
+      new_y
+      |> Nx.multiply(tangent)
+      |> then(&Nx.subtract(new_x, &1))
+      |> Nx.round()
+
+    new_y = Nx.subtract(new_ctr_height, new_y)
+    new_x = Nx.subtract(new_ctr_width, new_x)
+
+    new_y = Nx.as_type(Nx.round(new_y), {:u, 32})
+    new_x = Nx.as_type(Nx.round(new_x), {:u, 32})
+
+    Nx.stack([new_y, new_x], axis: 1)
+  end
+
+  defnp calculate_ctr(coordinate) do
+    coordinate
+    |> Nx.add(1)
+    |> Nx.divide(2)
+    |> Nx.subtract(1)
+    |> Nx.round()
+  end
+
+  defnp preprocess_position(pos, n) do
+    Nx.concatenate(
+      [
+        Nx.new_axis(pos, 1) |> Nx.tile([1, 4, 1]),
+        Nx.iota({n, 4, 1}, axis: 1)
+      ],
+      axis: 2
+    )
+    |> Nx.reshape({n * 4, 3})
+  end
+
+  # there is no pi const in Nx yet.
+  @pi 3.14159
+
+  # new_image is a tensor with dims of rotated image filled with zeros
+
+  defn rotate_image_by_angle(image, angle, new_image, radians? \\ 1) do
+    # if radians? is set to false than assuming that angle is given in degrees
+    angle = if radians? == 0, do: angle |> Nx.divide(180) |> Nx.multiply(@pi), else: angle
+
+    {height, width, _} = image.shape
+    {new_height, new_width, _} = new_image.shape
+
+    orig_ctr_height = calculate_ctr(height)
+    orig_ctr_width = calculate_ctr(width)
+
+    new_ctr_height = calculate_ctr(new_height)
+    new_ctr_width = calculate_ctr(new_width)
+
+    pos =
+      calculate_new_positions(
+        image.shape,
+        orig_ctr_height,
+        orig_ctr_width,
+        new_ctr_height,
+        new_ctr_width,
+        angle
+      )
+
+    {n, 2} = Nx.shape(pos)
+
+    preprocessed_pos = preprocess_position(pos, n)
+    image = image |> Nx.reshape({n * 4})
+
+    Nx.indexed_add(new_image, preprocessed_pos, image)
+  end
+end
+```
+
+## Example of usage
+
+```elixir
+{_, dir} = File.cwd()
+path_to_file = "#{dir}/<image>.png"
+path_to_save = "#{dir}/<image_rotated>.png"
+
+image = Preprocess.image_to_tensor(path_to_file)
+{height, width, _} = image.shape
+{new_height, new_width} = Preprocess.get_new_height_and_width(height, width, -75, 0)
+# 4 because there are 4 channels (3 BGR/RGB + 1 Alpha)
+new_image = Nx.broadcast(Nx.tensor([0], type: {:u, 8}), {new_height, new_width, 4})
+
+rotated_image = Rotate.rotate_image_by_angle(image, -75, new_image, 0)
+{_, rotated_image_mat} = OpenCV.Nx.to_mat(rotated_image)
+# convert to correct format, default for opencv is BGR
+OpenCV.imwrite(
```

Comment:
It would be neat to use [`Kino.Image`](https://hexdocs.pm/kino/Kino.Image.html) to render the image directly in Livebook.

Reply:
Oh, I somehow missed this comment, I will remove saving to the disk.